### PR TITLE
MINIFICPP-842 - CivetWeb tests should use random port

### PR DESCRIPTION
**Pull request by arpadboda**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: extensions/civetweb/processors/ListenHTTP.cpp [hunk 1]
```diff
@@ -27,11 +27,22 @@ namespace minifi {
 namespace processors {
 
 core::Property ListenHTTP::BasePath(
-    core::PropertyBuilder::createProperty("Base Path")->withDescription("Base path for incoming connections")->isRequired(false)->withDefaultValue<std::string>("contentListener")->build());
+    core::PropertyBuilder::createProperty("Base Path")->
```
Comment (phrocker) [inline extensions/civetweb/processors/ListenHTTP.cpp]: I know we don't quite yet lint extensions ( this will happen soon ) but I'm pretty sure this will need //NOLINT if you intend to keep it this way.

File: extensions/civetweb/processors/ListenHTTP.cpp [hunk 2]
```diff
@@ -27,11 +27,22 @@ namespace minifi {
...
-    core::PropertyBuilder::createProperty("Listening Port")->withDescription("The Port to listen on for incoming connections")->isRequired(true)->withDefaultValue<int>(
-        80, core::StandardValidators::PORT_VALIDATOR())->build());
+    core::PropertyBuilder::createProperty("Listening Port")->
+    withDescription("The Port to listen on for incoming connections")->
+    isRequired(true)->
+    withDefaultValue<int>(80, core::StandardValidators::PORT_VALIDATOR())->build());
+
+core::Property ListenHTTP::RandomPort(
+    core::PropertyBuilder::createProperty("Random Port")->
+    withDescription("Choose random port to listen - for testing purposes only!")->
```
Comment (phrocker) [inline extensions/civetweb/processors/ListenHTTP.cpp]: I don't think we should have testing code available to production environments. This will be relayed through the C2 framework. Perhaps eliminating it when not building tests will work.

File: extensions/civetweb/processors/ListenHTTP.cpp [hunk 3]
```diff
@@ -191,6 +211,16 @@ void ListenHTTP::onSchedule(core::ProcessContext *context, core::ProcessSessionF
   server_.reset(new CivetServer(options));
   handler_.reset(new Handler(basePath, context, sessionFactory, std::move(authDNPattern), std::move(headersAsAttributesPattern)));
   server_->addHandler(basePath, handler_.get());
+
+  if (randomPort) {
+    const auto& vec = server_->getListeningPorts();
+    if (vec.size() != 1) {
+      logger_->log_error("Random port is set, but there is no listening port! Server most probably failed to start!");
```
Comment (phrocker) [inline extensions/civetweb/processors/ListenHTTP.cpp]: If this is listening on http and https and a test will use either/both how will it know those random ports?

File: extensions/civetweb/processors/ListenHTTP.cpp [hunk 4]
```diff
@@ -441,6 +471,10 @@ int64_t ListenHTTP::WriteCallback::process(std::shared_ptr<io::BaseStream> strea
   return nlen;
 }
 
+std::string ListenHTTP::getPort() const {
+  return listeningPort;
```
Comment (phrocker) [inline extensions/civetweb/processors/ListenHTTP.cpp]: Why is a port a string? Is it because ports are appended with s?

File: extensions/civetweb/processors/ListenHTTP.cpp [hunk 5]
```diff
@@ -40,12 +51,18 @@ core::Property ListenHTTP::SSLCertificate("SSL Certificate", "File containing PE
 core::Property ListenHTTP::SSLCertificateAuthority("SSL Certificate Authority", "File containing trusted PEM-formatted certificates", "");
 
 core::Property ListenHTTP::SSLVerifyPeer(
-    core::PropertyBuilder::createProperty("SSL Verify Peer")->withDescription("Whether or not to verify the client's certificate (yes/no)")->isRequired(false)->withAllowableValue<std::string>("yes")
-        ->withAllowableValue("no")->withDefaultValue("no")->build());
+    core::PropertyBuilder::createProperty("SSL Verify Peer")->
+    withDescription("Whether or not to verify the client's certificate (yes/no)")->
```
Comment (phrocker) [inline extensions/civetweb/processors/ListenHTTP.cpp]: I'm not against this style change at all and the removal of the linted code, but let's move that to a separate PR so it can be tracked better through git.

Comment (phrocker) [reply extensions/civetweb/processors/ListenHTTP.cpp]: * sorry in the future. leave these as-is

File: extensions/civetweb/processors/ListenHTTP.cpp (hunk 3 above)
Comment (arpadboda) [reply extensions/civetweb/processors/ListenHTTP.cpp]: There is no such option. 
The processor currently only supports one port property, which is validated using port validator.
http XOR https depends on whether SSL cert property is provided or not. In case it is, the processor automatically switches to https. 
To make sure I will add a getter for this, too.

File: extensions/civetweb/processors/ListenHTTP.cpp (hunk 4 above)
Comment (arpadboda) [reply extensions/civetweb/processors/ListenHTTP.cpp]: Yes.

File: extensions/civetweb/processors/ListenHTTP.cpp (hunk 5 above)
Comment (arpadboda) [reply extensions/civetweb/processors/ListenHTTP.cpp]: Not "just" style, changed these to use allowableValue**s** to avoid calling _withAllowableValue_ many times for the same prop.

File: extensions/civetweb/processors/ListenHTTP.cpp (hunk 1 above)
Comment (arpadboda) [reply extensions/civetweb/processors/ListenHTTP.cpp]: That's a good point, thanks!
Checked other processors, it should be the opposite way ( -> operator is at the beginning of new lines) to be linter-compatible, so fixed.

File: extensions/civetweb/processors/ListenHTTP.cpp (hunk 2 above)
Comment (arpadboda) [reply extensions/civetweb/processors/ListenHTTP.cpp]: Good point. Added an ifdef to only add it to the set of supported properties in case of building with tests.

File: extensions/civetweb/processors/ListenHTTP.cpp (hunk 5 above)
Comment (phrocker) [reply extensions/civetweb/processors/ListenHTTP.cpp]: That is a fluent API.

**Comment by arpadboda** [comment]
> Was about to merge and realized that SKIP_TESTS is used as the ifdef. The logic there is that if tests are built include this parameter. This means that RandomPort will be included via the manifest if tests were built. Could you just set listening port to zero via the config of the test, use the function to get the port you added to later config the client? This way no randomport option is needed. The same port config is used.

That was my initial idea as well, although the port validator we have at the moment doesn't accept 0, so that wouldn't work. 
Which makes sense in a way as 0 is not a valid port. 

We could make that a global rule in MiNiFi - port validator allows 0 and that means the processor has to choose a random free port. We might consider this in the future, but it's a breaking change and clearly doesn't fit the scope of this ticket.
